x/stakeibc/keeper/msg_server_liquid_stake.go

```diff
@@ -17,114 +15,91 @@ import (
-	stAssetDenom := types.StAssetDenomFromHostZoneDenom(denom)
+	// The tokens that are sent to the protocol are denominated in the ibc hash of the native token on stride (e.g. ibc/xxx)
+	nativeDenom := hostZone.IbcDenom
+	nativeCoin := sdk.NewCoin(nativeDenom, msg.Amount)
+	if !types.IsIBCToken(nativeDenom) {
+		return nil, errorsmod.Wrapf(types.ErrInvalidToken, "denom is not an IBC token (%s)", nativeDenom)
+	}
 
-	// TODO(TEST-7): Add an exchange rate here! What object should we store the exchange rate on?
-	// How can we ensure that the exchange rate is not manipulated?
-	hz, _ := k.GetHostZoneFromHostDenom(ctx, denom)
-	amountToMint := (sdk.NewDecFromInt(amount).Quo(hz.RedemptionRate)).TruncateInt()
-	coinString := amountToMint.String() + stAssetDenom
-	stCoins, err := sdk.ParseCoinsNormalized(coinString)
-	if err != nil {
-		k.Logger(ctx).Error("Failed to parse coins")
-		return errorsmod.Wrapf(sdkerrors.ErrInvalidRequest, "Failed to parse coins %s", coinString)
+	// Confirm the user has a sufficient balance to execute the liquid stake
+	balance := k.bankKeeper.GetBalance(ctx, userAddress, nativeDenom)
+	if balance.IsLT(nativeCoin) {
+		return nil, errorsmod.Wrapf(sdkerrors.ErrInsufficientFunds, "balance is lower than staking amount. staking amount: %v, balance: %v", msg.Amount, balance.Amount)
 	}
 
-	// Mints coins to the module account, will error if the module account does not exist or is unauthorized.
+	// Determine the amount of stTokens to mint using the redemption rate
+	stAmount := (sdk.NewDecFromInt(msg.Amount).Quo(hostZone.RedemptionRate)).TruncateInt()
+	if stAmount.IsZero() {
+		return nil, errorsmod.Wrapf(types.ErrInsufficientLiquidStake,
+			"Liquid stake of %s%s would return 0 stTokens", msg.Amount.String(), hostZone.HostDenom)
+	}
 
-	err = k.bankKeeper.MintCoins(ctx, types.ModuleName, stCoins)
-	if err != nil {
-		k.Logger(ctx).Error("Failed to mint coins")
-		return errorsmod.Wrapf(sdkerrors.ErrInvalidRequest, "Failed to mint coins")
+	// Transfer the native tokens from the user to module account
+	if err := k.bankKeeper.SendCoins(ctx, userAddress, hostZoneAddress, sdk.NewCoins(nativeCoin)); err != nil {
+		return nil, errorsmod.Wrap(err, "failed to send tokens from Account to Module")
 	}
 
-	// transfer those coins to the user
-	err = k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, sender, stCoins)
-	if err != nil {
-		k.Logger(ctx).Error("Failed to send coins from module to account")
-		return errorsmod.Wrapf(sdkerrors.ErrInvalidRequest, "Failed to send %s from module to account", stCoins.GetDenomByIndex(0))
+	// Mint the stTokens and transfer them to the user
+	stDenom := types.StAssetDenomFromHostZoneDenom(msg.HostDenom)
+	stCoin := sdk.NewCoin(stDenom, stAmount)
+	if err := k.bankKeeper.MintCoins(ctx, types.ModuleName, sdk.NewCoins(stCoin)); err != nil {
+		return nil, errorsmod.Wrapf(err, "Failed to mint coins")
```

Comment:
Given this needs to work non-atomically - should we refund the coins sent in here if an error is thrown? I'd err on the side of no, but it will require a software upgrade to fix this if something goes wrong

Reply:
that's a good idea! Lemme read through the function again and get back to ya
Feels like there's other stuff we'd have to revert too (e.g. burn coins if the send from module account fails)

Reply:
chatted offline, I think we're in favor of leaving the function as is instead of creating a non-atomic version considering the low that the function fails with impartial state, and the higher probabillity of introducing a new bug by trying to add in reversal logic